### PR TITLE
XCM builder pattern improvement - Accept `impl Into<T>` instead of just `T`

### PR DESCRIPTION
**Pull request by franciscoaguirre**
The XCM builder pattern lets you build xcms like so:

```rust
let xcm = Xcm::builder()
    .withdraw_asset((Parent, 100u128).into())
    .buy_execution((Parent, 1u128).into())
    .deposit_asset(All.into(), AccountId32 { id: [0u8; 32], network: None }.into())
    .build();
```

All the `.into()` become quite annoying to have to write.
I accepted `impl Into<T>` instead of `T` in the generated methods from the macro.
Now the previous example can be simplified as follows:

```rust
let xcm = Xcm::builder()
    .withdraw_asset((Parent, 100u128))
    .buy_execution((Parent, 1u128))
    .deposit_asset(All, [0u8; 32])
    .build();
```

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5537759

File: polkadot/xcm/procedural/tests/builder_pattern.rs [hunk 1]
```diff
@@ -22,11 +22,11 @@ use xcm::latest::prelude::*;
 #[test]
 fn builder_pattern_works() {
 	let asset: Asset = (Here, 100u128).into();
-	let beneficiary: Location = AccountId32 { id: [0u8; 32], network: None }.into();
-	let message: Xcm<()> = Xcm::builder()
-		.receive_teleported_asset(asset.clone().into())
+	let beneficiary: Location = [0u8; 32].into();
+	let message: Xcm<()> = Xcm::builder_unsafe()
```
Comment (gupnik) [inline polkadot/xcm/procedural/tests/builder_pattern.rs:26]: Why did we change it to `builder_unsafe`?

Comment (franciscoaguirre) [reply polkadot/xcm/procedural/tests/builder_pattern.rs:26]: Was testing some stuff and didn't realize 😅, changing it back

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5543769

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5544065

**Review by gupnik — APPROVED** [review]
Should we add the builder pattern here as well? https://paritytech.github.io/xcm-docs/

**Review by bkchr — APPROVED** [review]

**Comment by acatangiu** [comment]
bot fmt

File: prdoc/pr_3708.prdoc [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+      let xcm = Xcm::builder()
+          .withdraw_asset((Parent, 100u128))
+          .buy_execution((Parent, 1u128))
+          .deposit_asset(All, [0u8; 32])
+          .build();
+      ```
+
+crates:
+- name: "xcm-procedural"
+- name: "staging-xcm"
```
Comment (acatangiu) [inline prdoc/pr_3708.prdoc:29]:
```suggestion
crates:
- name: "xcm-procedural"
  bump: minor
- name: "staging-xcm"
  bump: minor
```

**Review by acatangiu — APPROVED** [review]

**Comment by pgherveou** [comment]
just curious, @franciscoaguirre did you check whether or not this change impact the binary size ?

**Comment by franciscoaguirre** [comment]
> just curious, @franciscoaguirre did you check whether or not this change impact the binary size ?

No. Is it because of the `impl` that it might increase binary size much?

**Pull request merged**